lib/system/threads.nim

```diff
@@ -61,16 +61,42 @@ const
 # use ``stdcall`` since it is mapped to ``noconv`` on UNIX anyway.
 
 type
-  Thread*[TArg] = object
-    core: PGcThread
-    sys: SysThread
+  ThreadCore[TArg] = object
+    ## The internal managnement data associated with a thread. Allocated
+    ## by the spawning thread, and - initially - owned by both the spawning
+    ## and spawned thread (shared ownership). If the spawned thread finishes
+    ## before the `Thread <#Thread>` handle owned by the spawning thread goes
+    ## out of scope, the spawning thread frees the instance, otherwise the
+    ## spawned thread does.
+    when emulatedThreadVars:
+      tls: ThreadLocalStorage
+
+    rc: int
```

Comment:
Preface: I like pretty much everything about this PR. The only thing that's got me a little nervous is the reduction(?) of structure, although we didn't really have it because of crashes/undefined behaviour. Anyhow, this and likely all other comments in this PR are to see if we can have structured concurrency, or the infrastructure for it, or at least set it up so it's not too painful a transition in the future. With that in mind, my comment:

---

So, I'm wondering, if it might be better to have this:

```nim
  owner: ptr Thread
  alive: bool  ## whether the child is still alive
```

Together `owner` and `alive` would be equivalent to a ref count while also incorporating a single owner concept.

Future extension and usage: we could consider ownership change and ensuring a child doesn't exist longer than it should. This would also open up the possibility of setting a different intial `owner` in the future via an extended `createThread` api, that allowed spawning a thread by a child thread for a common parent thread. All the check related to this, would be runtime checks initially, but I bet we could statically check this stuff in the future.

Reply:
My view of `Thread` is that it's a low-level primitive that other facilities (like a thread pool) built upon.

My model of the current implementation:
* `Thread` is a handle to a thread. It has unique ownership of the thread. Moving the handle therefore moves ownership over it (to whoever owns the destination location)
* `ThreadCore` is the internal-only management data, invisible to the programmer. Its two possible owners are: the running thread and the corresponding `Thread` handle
* a *detached thread* is a thread that outlives the `Thread`

Structured concurrency facilities are then built upon `Thread` (at least that's how I think it could work). For example, a simple handle type that prevents the thread from outliving the handle could look like this:
```nim
type JThread[Arg] = object
  # the name is borrowed from C++'s `std::jthread`
  has: bool
  thread: Thread[Arg]

proc `=copy`[T](x: var JThread[T]) {.error.} # unique ownership

proc `=destroy`[T](x: var JThread[T]) =
  if x.has:
    x.thread.joinThread()
    `=destroy`(x.thread)

# when a `JThread` goes out of scope, the thread is awaited first 
```

---

Regarding the idea you brought up, I'm not sure I fully understand it:
* unique ownership for `ThreadCore` would mean that detaching threads is impossible
* I'm not sure what the knowledge of the exact owner is useful for. Re-parenting the internal management data of a thread to another `Thread` instance doesn't seem useful nor necessary to me
* `ThreadCore` keeping a reference to a `Thread` cannot work, because `Thread` is not guaranteed to be stationary in memory. If `Thread` moves, the pointer would become stale

> we could consider ownership change and ensuring a child doesn't exist longer than it should.

If I understand correctly, something like the `JThread` implementation from above could also achieve this?

> This would also open up the possibility of setting a different intial `owner` in the future via an extended `createThread` api, that allowed spawning a thread by a child thread for a common parent thread.

Hm, what would be the effective difference compared to:
```nim
var
  child: Thread[void]  

proc run() {.thread.} =
  # ---- child thread 1
  var t = (createThread[void])(proc () =
    # ---- child thread 2
    # do something
  )
  # ...
  # now move ownership of the child thread to the parent thread:
  {.cast(gcSafe).}:
    child = move t

# ---- main thread
var t = (createThread[void])(run)
t.joinThread() # wait for child thread 1 to finish
# the main thread now owns child thread 2
```

Reply:
Thank you for the thoughtful response, in hindsight I realized I made a pretty big error: I misread the diff thinking it was `Thread`, the `rc` field should have tipped me off, sorry.

> Structured concurrency facilities are then built upon `Thread` (at least that's how I think it could work). For example, a simple handle type that prevents the thread from outliving the handle could look like this:
> 
> ```nim
> type JThread[Arg] = object
>   # the name is borrowed from C++'s `std::jthread`
>   has: bool
>   thread: Thread[Arg]
> 
> proc `=copy`[T](x: var JThread[T]) {.error.} # unique ownership
> 
> proc `=destroy`[T](x: var JThread[T]) =
>   if x.has:
>     x.thread.joinThread()
>     `=destroy`(x.thread)
> 
> # when a `JThread` goes out of scope, the thread is awaited first 
> ```

Completely agree, the ownership concept should be in `Thread` or a derivative (e.g.: `jThread` as above). Yup, in a structured concurrency approach there would always have to be some "owner" or the thread is the root thread, akin to a "pid 1".

> > we could consider ownership change and ensuring a child doesn't exist longer than it should.
> 
> If I understand correctly, something like the `JThread` implementation from above could also achieve this?

Yes, although ideally we'd want to figure out/prevent threads escaping their "scope", my thinking was the thread resource itself was scoped, rather than solely focusing on the `Thread` handle. It would largely prevent accidentally moving a thread, via its handle `Thread`, to a location with a longer lifetime.

> > This would also open up the possibility of setting a different intial `owner` in the future via an extended `createThread` api, that allowed spawning a thread by a child thread for a common parent thread.
> 
> Hm, what would be the effective difference compared to:
> 
> ```nim
> var
>   child: Thread[void]  
> 
> proc run() {.thread.} =
>   # ---- child thread 1
>   var t = (createThread[void])(proc () =
>     # ---- child thread 2
>     # do something
>   )
>   # ...
>   # now move ownership of the child thread to the parent thread:
>   {.cast(gcSafe).}:
>     child = move t
> 
> # ---- main thread
> var t = (createThread[void])(run)
> t.joinThread() # wait for child thread 1 to finish
> # the main thread now owns child thread 2
> ```

It's similiar, but by having some sort of owner parameter we decouple location assignment/handling from the instantiator (the owner might want to change how/where things are stored).

This is firmly in the category of "potential future ideas": perhaps having `run` take a context parameter, which is then used in conjunction with `createThread` to create a `Thread` under `main`. At which point `createThread` would have to return a view, `lent Thread`, that could be used within `run`, but ultimately ownership of the viewed `Thread` lies with `main`. But I digress; and all this at least seems like a possibility we can **meaningfully** consider.

----

OK, having had a chance to read your comments and reply, this is my current thinking:

1. the current approach: at a minimum we a first pass runtime oriented structured concurrency approach
    1. this is categorically better than _none_, which is what we currently have
2. if we want more static enforcement, we probably need some facility, or convention, to pin a handle to some lifetime, such that intra-lifetime moves are ok, but not inter-ilfetime; but that's way out of scope (hah!) for this discussion

Reply:
Thanks a lot for elaborating, I think I understand better now the overall direction you want to take with structured concurrency.